plugins/single_plugins/preserve-output.cpp

```diff
@@ -0,0 +1,314 @@
+            outputs_being_removed.end(),
+            signal_data->output->to_string()), outputs_being_removed.end());
+        LOGD("Outputs currently being removed: ");
+        for (auto el : outputs_being_removed)
+        {
+            LOGD("- ", el);
+        }
+    };
+
+    wf::signal_connection_t restore_views_to_output = [=] (wf::signal_data_t *data)
+    {
+        auto signal_data = (wf::output_added_signal*)data;
+        if (signal_data->output != output)
+        {
+            // This event is not for this output
+            return;
+        }
+
+        std::string identifier = make_output_identifier(output);
+
+        // Focus this output if it was the last one focused
+        if (core_get_focused_output() == identifier)
+        {
+            LOGD("This is last focused output, refocusing: ", output->to_string());
+            wf::get_core().focus_output(output);
+            core_erase_focused_output();
+        }
+
+        for (auto& wo : wf::get_core().output_layout->get_outputs())
+        {
+            // Make a list of views to move to this output
+            auto views = std::vector<wayfire_view>();
+            for (auto& view : wo->workspace->get_views_in_layer(wf::LAYER_WORKSPACE))
+            {
+                if ((view->role != wf::VIEW_ROLE_TOPLEVEL) || !view->is_mapped())
+                {
+                    continue;
+                }
+
+                if (!view_has_data(view))
+                {
+                    continue;
+                }
+
+                auto last_output_info = view_get_data(view);
+                if (last_output_info->output_identifier == identifier)
+                {
+                    views.push_back(view);
+                }
+            }
+
+            // Sorts with the views closest to front last
+            std::sort(views.begin(), views.end(), last_output_info_z_sort());
+
+            // Move views to this output
+            for (auto& view : views)
+            {
+                auto last_output_info = view_get_data(view);
+                LOGD("Restoring view: ",
+                    view->get_title(), " to: ", output->to_string());
+
+                wf::get_core().move_view_to_output(view, output, false);
+                view->set_fullscreen(last_output_info->fullscreen);
+                view->set_minimized(last_output_info->minimized);
+                view->set_sticky(last_output_info->sticky);
+                view->set_tiled(last_output_info->tiled_edges);
+                if (last_output_info->tiled_edges != 0)
+                {
+                    view->tile_request(last_output_info->tiled_edges);
+                }
+
+                // This is to ensure the next resize is actually done, helps ensure
+                // apps redraw themselves too
+                view->resize(last_output_info->geometry.width + 1,
+                    last_output_info->geometry.height);
+                view->set_geometry(last_output_info->geometry);
```

Comment:
This sounds more like a bug in core if you need to write such hacks ... How can I reproduce the bug you're working around here?

Reply:
There's code I think the XDG view resize code that checks what the last resize sent was, and doesn't send the resize if so. Possibly this code is getting out of sync with the actual state of the XDG view, I'll try to get more info later on.

Reply:
In `view-impl.hpp` we have: 
```
wf::dimensions_t last_size_request
```
The `should_resize_client` method uses it to decide if we should send the request, and we only send it if the new request is different to the previous request. For some reason, this is stopping the view->set_geometry in my plugin from actually sending the resize request - and it needs to be sent because the views have been shrunk after their move to the (usually smaller) NOOP-1 output.

I guess this means that moving the views to NOOP-1 causes the views' size to be reduced somehow without setting this last_size_request - which is odd because I thought the view resizing after moving a view to a smaller output used the view::resize method.

Probably the easiest way to reproduce this yourself is to comment out the view->resize line in this code and try it yourself, if your normal screens are bigger than the NOOP it should reproduce it.

Reply:
@ammen99 I've been trying to get some debug info to shed some light on this, unfortunately it seems in the last week or so a commit to wlroots has broken DPMS for me - my monitors now don't switch back on after trying to wake them up and I have to reboot. I'll raise a wlroots bug for this, in the meantime are there any other changes you want to the code?

Reply:
Hi, unfortunately, I cannot reproduce even though I tried. However, I have a hunch what might be happening, because I have debugged (and thought I had fixed) a very similar bug in the past.

The basic problem is this: we have APIs for just moving, just resizing or moving+resizing a view. Now, a lot of places are using the third operation (`set_geometry()`) when they only need to move the view.

A more detailed explanation why this is problematic: when moving a view with `set_geometry` the usual computation is:

```cpp
auto wm = view->get_wm_geometry();
wm.x += 5;
wm.y += 5;
view->set_wm_geometry();
```

Now consider what would happen if the view had size 100x100, your plugin tries to resize it to 50x50: the `last_size_request` becomes 50x50, but the wm_geometry is still 100x100 because the client has not updated its state.
Now a plugin calls the thing from above, and the view is resized back to 100x100 !

Now, onto the interesting part: how do we fix it? We can't really do something like I have done here: http://ix.io/2JbQ (though that patch might be useful for testing if my hypothesis is correct), because there are cases where two plugins actually compete for setting the view geometry and the second one should win (this for example of simple-tile, which always keeps its views in a fixed layout/geometries). Instead, we should find all incorrect usages of `set_geometry` and replace them with `move`. I'd have done this myself if I could reproduce, but I cannot, so I ask you to do this - print the geometry in `set_geometry`, and also use `wf::print_trace(false/true)` (false is slower, but usually gives more info) to see where set_geometry is called, and to see what the race is.

Reply:
@ammen99 - I believe I've figured it out. The relevant method is `wayfire_xdg_view::resize`.

First, the view is resized when it's moved to the NOOP. This resize is sent to the client (`wlr_xdg_toplevel_set_size`), because the new size is different to the client's current_size (from `wlr_xdg_surface_get_geometry`).

Second, the preserve-output plugin requests the view be resized back to what it was. It seems the client app (or wlroots?) hasn't actually processed the previous resize, because it's actual current_size hasn't changed yet. So this resize is ignored.

Reply:
The resize does not take effect immediately, the configure request is sent on the next idle, and then the client needs to submit a new buffer. Only after that we have the new size. Anyway, I don't understand why the second resize is ignored?

Reply:
1. The view has size 1885x1025
2. During DPMS, the view is moved to NOOP, a resize request is sent to resize it down to 1178x683
3. The preserve-output plugin calls resize to set it to 1885x1025
4. Because the NOOP's resize request in step 2 hasn't been processed yet, `wlr_xdg_surface_get_geometry` returns 1885x1025 for the view - this means the resize call to set it to 1885x1025 isn't sent

In `wayfire_xdg_view::resize` we have this code checking if we should send the resize before sending it:
```
    wf::dimensions_t current_size{current_geometry.width, current_geometry.height};                                    
    if (should_resize_client({w, h}, current_size))                                    
    {                                                                                                          
        this->last_size_request = {w, h};                                                                   
        wlr_xdg_toplevel_set_size(xdg_toplevel->base, w, h);                                                                                                                                   
    } 
```
`should_resize_client` checks the resize request isn't the same as the last resize request sent to the client, and isn't the same as the current size of the client (current_size).

Reply:
Maybe `should_resize_client` should be changed to do the resize if the new size is different to the last size request *or* the current size? Rather than needing to be different to both as I think is the case now.

Reply:
Nevermind I misread the should_resize_client method, will dig into it more tomorrow.

Reply:
Ok, I'm lost now. I've added this debug into the resize method:
```
void wayfire_xdg_view::resize(int w, int h)                                    
{                                    
    LOGD(this->get_title(), " resize called: ", w, "x", h);                                    
    LOGD("This ref: ", this);                                    
    wf::print_trace(true);                                    
    if (view_impl->frame)                                    
    {                                    
        view_impl->frame->calculate_resize_size(w, h);                                    
    }                                    
                                    
    auto current_geometry = get_xdg_geometry(xdg_toplevel);                                    
    wf::dimensions_t current_size{current_geometry.width, current_geometry.height};                                    
    LOGD("- Current size: ", current_size.width, "x", current_size.height);                                    
    LOGD("- last size request: ", this->last_size_request.width, "x", this->last_size_request.height);                                    
    if (should_resize_client({w, h}, current_size))                                    
    {                                    
        LOGD("- YES - Doing resize!");                                    
        this->last_size_request = {w, h};                                    
        LOGD("- NEW last size request: ", this->last_size_request.width, "x", this->last_size_request.height);                                                                                 
        wlr_xdg_toplevel_set_size(xdg_toplevel->base, w, h);                                    
    }                                    
    else {                                    
        LOGD("- NO - Not doing resize.");                                    
    }                                    
}                                    
```

And this is what we get in the logs:
```
 769 DD 24-12-20 11:17:28.073 - [src/view/xdg-shell.cpp:394] Volume Control resize called: 1178x683                                                                                            
 770 DD 24-12-20 11:17:28.073 - [src/view/xdg-shell.cpp:395] This ref: 0x55c45474cef0
 771 EE 24-12-20 11:17:28.078 - #1  wayfire_xdg_view::resize(int, int) at 0x55c451486595
 772 EE 24-12-20 11:17:28.079 - #2  wf::view_interface_t::set_geometry(wlr_box) at 0x55c4514703c8
 773 EE 24-12-20 11:17:28.079 - #3  wf::compositor_core_impl_t::move_view_to_output(nonstd::observer_ptr<wf::view_interface_t>, wf::output_t*, bool) at 0x55c451406ea4
 774 EE 24-12-20 11:17:28.079 - #4  wf::transfer_views(wf::output_t*, wf::output_t*) at 0x55c4513cbf08
 775 EE 24-12-20 11:17:28.079 - #5  wf::output_layout_output_t::destroy_wayfire_output(bool) at 0x55c4513ce97f
 776 EE 24-12-20 11:17:28.079 - #6  wf::output_layout_output_t::apply_state(wf::output_state_t const&, bool) at 0x55c4513d00eb
 777 EE 24-12-20 11:17:28.079 - #7  wf::output_layout_t::impl::apply_configuration(std::map<wlr_output*, wf::output_state_t, std::less<wlr_output*>, std::allocator<std::pair<wlr_output* const     , wf::output_state_t> > > const&) at 0x55c4513d2e57
 778 EE 24-12-20 11:17:28.079 - #8  wf::output_layout_t::impl::remove_output(wlr_output*) at 0x55c4513d1821
 779 EE 24-12-20 11:17:28.079 - #9  wf::output_layout_t::impl::add_output(wlr_output*)::{lambda(void*)#1}::operator()(void*) const at 0x55c4513d14d4
 780 EE 24-12-20 11:17:28.079 - #10 void std::__invoke_impl<void, wf::output_layout_t::impl::add_output(wlr_output*)::{lambda(void*)#1}&, void*>(std::__invoke_other, wf::output_layout_t::impl     ::add_output(wlr_output*)::{lambda(void*)#1}&, void*&&) at 0x55c4513e0f14
 781 EE 24-12-20 11:17:28.079 - #11 std::enable_if<is_invocable_r_v<void, wf::output_layout_t::impl::add_output(wlr_output*)::{lambda(void*)#1}&, void*>, std::enable_if>::type std::__invoke_r     <void, wf::output_layout_t::impl::add_output(wlr_output*)::{lambda(void*)#1}&, void*>(void&&, (wf::output_layout_t::impl::add_output(wlr_output*)::{lambda(void*)#1}&)...) at 0x55c4513dd9     4c
 782 EE 24-12-20 11:17:28.079 - #12 std::_Function_handler<void (void*), wf::output_layout_t::impl::add_output(wlr_output*)::{lambda(void*)#1}>::_M_invoke(std::_Any_data const&, void*&&) at 0     x55c4513d9ca0                                                                                   
 783 EE 24-12-20 11:17:28.079 - #13 std::function<void (void*)>::operator()(void*) const at 0x55c4513cabe1
 784 EE 24-12-20 11:17:28.079 - #14 wf::wl_listener_wrapper::emit(void*) at 0x55c4513ca5a1                    
 785 EE 24-12-20 11:17:28.079 - #15 /home/jordan/GitRepo/wayfire-fork/build/src/wayfire(+0x2203d2) [0x55c4513ca3d2]
 786 EE 24-12-20 11:17:28.079 - #16 /usr/lib/libwlroots.so.7(+0x8eab7) [0x7f0506230ab7]          
 787 EE 24-12-20 11:17:28.079 - #17 /usr/lib/libwlroots.so.7(wlr_output_destroy+0x4f) [0x7f050621d614]
 788 EE 24-12-20 11:17:28.079 - #18 /usr/lib/libwlroots.so.7(+0x37c40) [0x7f05061d9c40]
 789 EE 24-12-20 11:17:28.079 - #19 /usr/lib/libwlroots.so.7(+0x37407) [0x7f05061d9407]
 790 EE 24-12-20 11:17:28.079 - #20 /usr/lib/libwlroots.so.7(+0x3227a) [0x7f05061d427a]
 791 EE 24-12-20 11:17:28.079 - #21 /usr/lib/libwlroots.so.7(+0x8eab7) [0x7f0506230ab7]
 792 EE 24-12-20 11:17:28.079 - #22 /usr/lib/libwlroots.so.7(+0x4e751) [0x7f05061f0751]                     
 793 EE 24-12-20 11:17:28.079 - #23 /usr/lib/libwayland-server.so.0(wl_event_loop_dispatch+0xca) [0x7f0506286faa]
 794 EE 24-12-20 11:17:28.080 - #24 /usr/lib/libwayland-server.so.0(wl_display_run+0x27) [0x7f05062854e7]      
 795 EE 24-12-20 11:17:28.080 - #25 /home/jordan/GitRepo/wayfire-fork/build/src/wayfire(main+0xc87) [0x55c4513c1d34]
 796 EE 24-12-20 11:17:28.080 - #26 /usr/lib/libc.so.6(__libc_start_main+0xf2) [0x7f0505945152]                 
 797 EE 24-12-20 11:17:28.080 - #27 /home/jordan/GitRepo/wayfire-fork/build/src/wayfire(_start+0x2e) [0x55c4513be70e]
 798 DD 24-12-20 11:17:28.080 - [src/view/xdg-shell.cpp:404] - Current size: 1885x1025
 799 DD 24-12-20 11:17:28.080 - [src/view/xdg-shell.cpp:405] - last size request: 1885x1025
 800 DD 24-12-20 11:17:28.080 - [src/view/xdg-shell.cpp:408] - YES - Doing resize!       
 801 DD 24-12-20 11:17:28.080 - [src/view/xdg-shell.cpp:410] - NEW last size request: 1178x683                 

(SNIP)

1102 DD 24-12-20 11:17:28.940 - [src/view/xdg-shell.cpp:394] Volume Control resize called: 1885x1025
1103 DD 24-12-20 11:17:28.940 - [src/view/xdg-shell.cpp:395] This ref: 0x55c45474cef0
1104 EE 24-12-20 11:17:28.941 - #1  wayfire_xdg_view::resize(int, int) at 0x55c451486595
1105 EE 24-12-20 11:17:28.941 - #2  wf::view_interface_t::set_geometry(wlr_box) at 0x55c4514703c8
1106 EE 24-12-20 11:17:28.941 - #3  wayfire_preserve_output::restore_views_to_output() at 0x7f04c0f64195
1107 EE 24-12-20 11:17:28.941 - #4  wayfire_preserve_output::init()::{lambda()#1}::operator()() const at 0x7f04c0f64513
1108 EE 24-12-20 11:17:28.941 - #5  void std::__invoke_impl<void, wayfire_preserve_output::init()::{lambda()#1}&>(std::__invoke_other, wayfire_preserve_output::init()::{lambda()#1}&) at 0x7f0     4c0f699e4
1109 EE 24-12-20 11:17:28.941 - #6  std::enable_if<is_invocable_r_v<void, wayfire_preserve_output::init()::{lambda()#1}&>, std::enable_if>::type std::__invoke_r<void, wayfire_preserve_output:     :init()::{lambda()#1}&>(void&&, (wayfire_preserve_output::init()::{lambda()#1}&)...) at 0x7f04c0f68865
1110 EE 24-12-20 11:17:28.941 - #7  std::_Function_handler<void (), wayfire_preserve_output::init()::{lambda()#1}>::_M_invoke(std::_Any_data const&) at 0x7f04c0f67703
1111 EE 24-12-20 11:17:28.941 - #8  std::function<void ()>::operator()() const at 0x55c4513c3d30
1112 EE 24-12-20 11:17:28.941 - #9  wf::wl_idle_call::execute() at 0x55c4513ca7fe
1113 EE 24-12-20 11:17:28.941 - #10 /home/jordan/GitRepo/wayfire-fork/build/src/wayfire(+0x2203f5) [0x55c4513ca3f5]
1114 EE 24-12-20 11:17:28.941 - #11 /usr/lib/libwayland-server.so.0(wl_event_loop_dispatch_idle+0x1b) [0x7f0506286ebb]
1115 EE 24-12-20 11:17:28.941 - #12 /usr/lib/libwayland-server.so.0(wl_event_loop_dispatch+0xeb) [0x7f0506286fcb]
1116 EE 24-12-20 11:17:28.941 - #13 /usr/lib/libwayland-server.so.0(wl_display_run+0x27) [0x7f05062854e7]
1117 EE 24-12-20 11:17:28.941 - #14 /home/jordan/GitRepo/wayfire-fork/build/src/wayfire(main+0xc87) [0x55c4513c1d34]
1118 EE 24-12-20 11:17:28.941 - #15 /usr/lib/libc.so.6(__libc_start_main+0xf2) [0x7f0505945152]
1119 EE 24-12-20 11:17:28.941 - #16 /home/jordan/GitRepo/wayfire-fork/build/src/wayfire(_start+0x2e) [0x55c4513be70e]
1120 DD 24-12-20 11:17:28.941 - [src/view/xdg-shell.cpp:404] - Current size: 1885x1025
1121 DD 24-12-20 11:17:28.941 - [src/view/xdg-shell.cpp:405] - last size request: 1885x1025
1122 DD 24-12-20 11:17:28.941 - [src/view/xdg-shell.cpp:414] - NO - Not doing resize.
```

`last_resize_request` is being changed back after it's set by the moving-to-NOOP resize. And there's no other instances of this resize method being called for this view in the log other than these two.

The only other place in the whole codebase where last_size_request is being set that could possibly make something like this happen is in `wayfire_xdg_view::commit`:
```
    auto xdg_g = get_xdg_geometry(xdg_toplevel);
(...)
    this->last_size_request = wf::dimensions(xdg_g);
```
So, it has to be that method causing this somehow.

Reply:
After the NOOP resize, this is the trace of the call to commit that is changing the last_size_request back, and thus stopping the second resize from working:
```
EE 24-12-20 12:17:32.409 - #1  wayfire_xdg_view::commit() ../src/view/xdg-shell.cpp:337                                                                                        
EE 24-12-20 12:17:32.456 - #2  wf::wlr_surface_base_t::wlr_surface_base_t(wf::surface_interface_t*)::{lambda(void*)#2}::operator()(void*) const ../src/view/surface.cpp:237    
EE 24-12-20 12:17:32.523 - #3  void std::__invoke_impl<void, wf::wlr_surface_base_t::wlr_surface_base_t(wf::surface_interface_t*)::{lambda(void*)#2}&, void*>(std::__invoke_other, wf::wlr_surface_base_t::wlr_surface_base_t(wf::surface_interface_t*)::{lambda(void*)#2}&, void*&&) /usr/include/c++/10.2.0/bits/invoke.h:60        
EE 24-12-20 12:17:32.588 - #4  ...) /usr/include/c++/10.2.0/bits/invoke.h:116    
EE 24-12-20 12:17:32.641 - #5  std::_Function_handler<void (void*), wf::wlr_surface_base_t::wlr_surface_base_t(wf::surface_interface_t*)::{lambda(void*)#2}>::_M_invoke(std::_Any_data const&, void*&&) /usr/include/c++/10.2.0/bits/std_function.h:292                                                                               
EE 24-12-20 12:17:32.677 - #6  std::function<void (void*)>::operator()(void*) const /usr/include/c++/10.2.0/bits/std_function.h:622    
EE 24-12-20 12:17:32.715 - #7  wf::wl_listener_wrapper::emit(void*) ../src/util.cpp:524                                         
EE 24-12-20 12:17:32.753 - #8  handle_wrapped_listener(wl_listener*, void*) ../src/util.cpp:460                                                                        
EE 24-12-20 12:17:32.771 - #9  wlr_signal_emit_safe ../util/signal.c:22                                                    
EE 24-12-20 12:17:32.789 - #10 surface_commit_pending ../types/wlr_surface.c:431                                                   
EE 24-12-20 12:17:32.807 - #11 surface_commit ../types/wlr_surface.c:502                                              
EE 24-12-20 12:17:32.819 - #12 ffi_prep_go_closure ??:?                                                               
EE 24-12-20 12:17:32.836 - #13 ffi_closure_free ??:?                                                                  
EE 24-12-20 12:17:32.854 - #14 wl_event_loop_get_destroy_listener ??:?                                                
EE 24-12-20 12:17:32.876 - #15 wl_client_destroy ??:?                                                                 
EE 24-12-20 12:17:32.895 - #16 wl_event_loop_dispatch ??:?                                                            
EE 24-12-20 12:17:32.907 - #17 wl_display_run ??:?                                                                    
EE 24-12-20 12:17:32.942 - #18 main ../src/main.cpp:397                                                               
EE 24-12-20 12:17:32.964 - #19 __libc_start_main ??:?                                                                 
EE 24-12-20 12:17:33.011 - #20 _start ??:?                                                                  
```

Reply:
Could you try with this patch here: http://ix.io/2Joj ?

Reply:
@ammen99 That has fixed the issue :)

Reply:
Are you going to commit that to master, or should I add it as a commit to this PR?

Reply:
Thanks for testing! I'll open a PR with it. Sadly, this will not work for Xwayland clients, because there we don't have any serial we can check against ...

Reply:
#973 has these changes, and also a detailed explanation of what goes wrong in case you're interested in the details :)

Reply:
Just tested with Minecraft and Steam, this issue doesn't affect Xwayland clients anyway!

Reply:
> Thanks for testing! I'll open a PR with it. Sadly, this will not work for Xwayland clients, because there we don't have any serial we can check against ...

FWIW there is _NET_WM_USER_TIME

Reply:
_NET_WM_USER_TIME is something different from xdg-shell serials. xdg-shell serial is used to count the requests from the compositor to the view, user-time is a timestamp on the last activity, for ex. key press or pointer motion.

Reply:
> Just tested with Minecraft and Steam, this issue doesn't affect Xwayland clients anyway!

Hmm, maybe the race doesn't happen there.. But I wouldn't complain if it works :D